onnx/defs/generator/defs.cc

```diff
@@ -131,7 +132,7 @@ TensorProto message.
     .Input(
            0,
            "input",
-           "Input tensor to provide shape information.", "tensor(int32)")
+           "Input tensor to copy shape and optionally type information from.", "T")
```

Comment:
Input and output types should not use the same parameter, they can be different.

Reply:
so the out type inference will be, if the "dtype" specified, then output type is "dtype", otherwise, it will use the same type as input?

Reply:
yes, that's my understanding too. so, we should use T1 and T2 for input and output type.

Reply:
What should the input type constraints be?

Reply:
Good question! I guess we have two options:
(1) Allow only the same types permitted for output, or
(2) Allow any type, and have the type-checker separately check that the input type is one of the permitted output types if "dtype" is missing.
(2) seems the right solution since it is less restrictive. (But it is a bit extra work.)

Reply:
Vote for (2)